components/org.wso2.carbon.identity.api.user.totp/org.wso2.carbon.identity.api.user.totp.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/totp/v1/core/TOTPService.java

```diff
@@ -0,0 +1,268 @@
+import org.wso2.carbon.context.PrivilegedCarbonContext;
+import org.wso2.carbon.core.util.CryptoException;
+import org.wso2.carbon.identity.api.user.common.error.APIError;
+import org.wso2.carbon.identity.api.user.common.error.ErrorResponse;
+import org.wso2.carbon.identity.api.user.totp.common.TOTPConstants;
+import org.wso2.carbon.identity.application.authentication.framework.exception.AuthenticationFailedException;
+import org.wso2.carbon.identity.application.authentication.framework.model.AuthenticatedUser;
+import org.wso2.carbon.identity.application.authenticator.totp.TOTPAuthenticatorConstants;
+import org.wso2.carbon.identity.application.authenticator.totp.TOTPKeyGenerator;
+import org.wso2.carbon.identity.application.authenticator.totp.exception.TOTPException;
+import org.wso2.carbon.identity.application.authenticator.totp.util.TOTPAuthenticatorConfig;
+import org.wso2.carbon.identity.application.authenticator.totp.util.TOTPAuthenticatorCredentials;
+import org.wso2.carbon.identity.application.authenticator.totp.util.TOTPAuthenticatorKey;
+import org.wso2.carbon.identity.application.authenticator.totp.util.TOTPKeyRepresentation;
+import org.wso2.carbon.identity.application.authenticator.totp.util.TOTPUtil;
+import org.wso2.carbon.identity.rest.api.user.totp.v1.dto.TOTPResponseDTO;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.utils.multitenancy.MultitenantUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.ws.rs.core.Response;
+
+import static org.wso2.carbon.identity.api.user.totp.common.TOTPConstants.ErrorMessage.SERVER_ERROR_DECRYPTING_SECRET;
+import static org.wso2.carbon.identity.api.user.totp.common.TOTPConstants.ErrorMessage.SERVER_ERROR_GENERAL;
+import static org.wso2.carbon.identity.api.user.totp.common.TOTPConstants.ErrorMessage
+        .SERVER_ERROR_RETRIEVING_REALM_FOR_USER;
+import static org.wso2.carbon.identity.api.user.totp.common.TOTPConstants.ErrorMessage
+        .SERVER_ERROR_RETRIEVING_USERSTORE_MANAGER;
+import static org.wso2.carbon.identity.api.user.totp.common.TOTPConstants.ErrorMessage.USER_ERROR_UNAUTHORIZED_USER;
+
+/**
+ * Service class for TOTP API
+ */
+public class TOTPService {
+
+    private static final Log log = LogFactory.getLog(TOTPService.class);
+
+    /**
+     * Retrieve the secret key of a given user.
+     *
+     * @return Secret Key.
+     */
+    public TOTPResponseDTO getSecretKey() {
+
+        UserRealm userRealm;
+        String encoding;
+
+        AuthenticatedUser user = getUser();
+        TOTPResponseDTO totpResponseDTO = new TOTPResponseDTO();
+        String secretKey = null;
+        String tenantAwareUsername = null;
+        Map<String, String> claims = new HashMap<>();
+
+        try {
+            userRealm = TOTPUtil.getUserRealm(user.toString());
+            tenantAwareUsername = MultitenantUtils.getTenantAwareUsername(user.toString());
```

Comment:
Same as above, I think it is better if we can use get username or similar method.